File: go/internal/dump/dump.go
// Package dump produces the canonical model dump — the cross-language
// conformance contract (conformance/README.md). Output matches
// JSON.stringify(value, null, 1): 1-space indent, fixed key order, shortest
// round-trip numbers, a trailing newline, UTF-8.
package dump

import (
	"sort"
	"strconv"
	"strings"

	"github.com/lprhodes/grid-md/go/internal/model"
	"github.com/lprhodes/grid-md/go/internal/numfmt"
	"github.com/lprhodes/grid-md/go/internal/refs"
	"github.com/lprhodes/grid-md/go/internal/scalar"
)

// omap is an insertion-ordered object for the dump tree.
type omap struct {
	keys []string
	vals []interface{}
}

func newOMap() *omap { return &omap{} }

func (m *omap) set(k string, v interface{}) {
	m.keys = append(m.keys, k)
	m.vals = append(m.vals, v)
}

// Model renders the canonical dump for a materialized workbook.
func Model(wb *model.Workbook) string {
	fm := wb.FM
	out := newOMap()
	out.set("gridmd", scalarLike(fm["gridmd"]))
	out.set("title", scalarLike(fm["title"]))
	dateSystem := 1900.0
	if ds, ok := fm["date-system"].(float64); ok && ds == 1904 {
		dateSystem = 1904
	}
	out.set("dateSystem", dateSystem)
	out.set("names", dumpNames(fm))

	sheets := make([]interface{}, 0, len(wb.Sheets))
	for _, s := range wb.Sheets {
		sheets = append(sheets, dumpSheet(s))
	}
	out.set("sheets", sheets)

	var b strings.Builder
	serialize(&b, out, 0)
	b.WriteByte('\n')
	return b.String()
}

func dumpNames(fm map[string]interface{}) []interface{} {
	raw, _ := fm["names"].([]interface{})
	type nameEntry struct {
		name string
		om   *omap
	}
	var entries []nameEntry
	for _, item := range raw {
		n, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		name := asString(n["name"])
		om := newOMap()
		om.set("name", name)
		om.set("ref", refOrNil(n, "ref"))
		om.set("formula", refOrNil(n, "formula"))
		if v, ok := n["value"]; ok {
			om.set("value", stringify(v))
		} else {
			om.set("value", nil)
		}
		entries = append(entries, nameEntry{name: name, om: om})
	}
	sort.SliceStable(entries, func(i, j int) bool { return entries[i].name < entries[j].name })
	out := make([]interface{}, 0, len(entries))
	for _, e := range entries {
		out = append(out, e.om)
	}
	return out
}

func refOrNil(m map[string]interface{}, key string) interface{} {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	return scalarLike(v)
}

func dumpSheet(s *model.Sheet) *omap {
	om := newOMap()
	om.set("name", s.Name)
	om.set("kind", s.Kind)
	om.set("hidden", hiddenValue(s.Meta["hidden"]))
	if fz, ok := s.Meta["freeze"].(string); ok {
		om.set("freeze", fz)
	} else {
		om.set("freeze", nil)
	}
	protected := false
	if pm, ok := s.Meta["protect"].(map[string]interface{}); ok {
		protected = boolish(pm["enabled"])
	}
	om.set("protected", protected)
	om.set("cells", dumpCells(s))
	om.set("merges", dumpMerges(s))
	om.set("tables", dumpTables(s))

	counts := newOMap()
	counts.set("cf", float64(s.CFRules))
	counts.set("validations", float64(s.Validations))
	counts.set("notes", float64(s.Notes))
	counts.set("threads", float64(s.Threads))
	counts.set("scenarios", float64(s.Scenarios))
	counts.set("sparklines", float64(s.Sparklines))
	counts.set("charts", float64(s.Charts))
	counts.set("pivots", float64(s.Pivots))
	counts.set("slicers", float64(s.Slicers))
	counts.set("images", float64(s.Images))
	counts.set("shapes", float64(s.Shapes))
	counts.set("hyperlinks", float64(s.Hyperlinks))
	om.set("counts", counts)
	return om
}

func dumpCells(s *model.Sheet) *omap {
	cells := s.Cells()
	sorted := make([]*model.Cell, len(cells))
	copy(sorted, cells)
	sort.SliceStable(sorted, func(i, j int) bool {
		if sorted[i].Row != sorted[j].Row {
			return sorted[i].Row < sorted[j].Row
		}
		return sorted[i].Col < sorted[j].Col
	})
	om := newOMap()
	for _, c := range sorted {
		// Every materialized cell is content-bearing (model.setContent), so no
		// nil-content filter is needed here.
		ref := refs.NumToCol(c.Col) + strconv.Itoa(c.Row)
		om.set(ref, dumpContent(c.Content))
	}
	return om
}

func dumpContent(ct *model.Content) interface{} {
	if ct.Rich != nil {
		var b strings.Builder
		for _, run := range ct.Rich {
			b.WriteString(asString(run["text"]))
		}
		e := newOMap()
		e.set("t", "rich")
		e.set("v", b.String())
		return e
	}
	if ct.Formula != nil {
		e := newOMap()
		e.set("t", "f")
		e.set("f", *ct.Formula)
		e.set("cached", scalarDump(ct.Cached))
		if ct.ArrayRef != nil {
			e.set("array", *ct.ArrayRef)
		} else {
			e.set("array", nil)
		}
		return e
	}
	return scalarDump(ct.Scalar)
}

func scalarDump(s *scalar.Scalar) interface{} {
	if s == nil {
		return nil
	}
	e := newOMap()
	switch s.Kind {
	case scalar.Number:
		e.set("t", "n")
		e.set("v", s.Num)
	case scalar.Boolean:
		e.set("t", "b")
		e.set("v", s.Bool)
	case scalar.Error:
		e.set("t", "e")
		e.set("v", s.Str)
	case scalar.Date, scalar.Time:
		e.set("t", "d")
		e.set("v", s.Str)
	default:
		e.set("t", "s")
		e.set("v", s.Str)
	}
	return e
}

func dumpMerges(s *model.Sheet) []interface{} {
	refsList := make([]string, 0, len(s.Merges))
	for _, m := range s.Merges {
		refsList = append(refsList, refs.NumToCol(m.C1)+strconv.Itoa(m.R1)+":"+refs.NumToCol(m.C2)+strconv.Itoa(m.R2))
	}
	sort.Strings(refsList)
	out := make([]interface{}, 0, len(refsList))
	for _, r := range refsList {
		out = append(out, r)
	}
	return out
}

func dumpTables(s *model.Sheet) []interface{} {
	sorted := make([]model.Table, len(s.Tables))
	copy(sorted, s.Tables)
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].Name < sorted[j].Name })
	out := make([]interface{}, 0, len(sorted))
	for _, t := range sorted {
		om := newOMap()
		om.set("name", t.Name)
		om.set("anchor", refs.NumToCol(t.Anchor.Col)+strconv.Itoa(t.Anchor.Row))
		cols := make([]interface{}, 0, len(t.Columns))
		for _, c := range t.Columns {
			cols = append(cols, c)
		}
		om.set("columns", cols)
		om.set("bodyRows", float64(t.BodyRows))
		om.set("hasTotals", t.HasTotals)
		out = append(out, om)
	}
	return out
}

func hiddenValue(v interface{}) interface{} {
	if v == true {
		return true
	}
	if v == "very" {
		return "very"
	}
	return false
}

func boolish(v interface{}) bool {
	b, ok := v.(bool)
	return ok && b
}

// scalarLike passes through the primitive JSON types; anything else becomes nil.
func scalarLike(v interface{}) interface{} {
	switch v.(type) {
	case nil, string, bool, float64:
		return v
	default:
		return nil
	}
}

func asString(v interface{}) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

// stringify mirrors JS String(v) for the YAML subset value types.
func stringify(v interface{}) interface{} {
	switch val := v.(type) {
	case string:
		return val
	case bool:
		if val {
			return "true"
		}
		return "false"
	case float64:
		return numfmt.Format(val)
	case nil:
		return ""
	default:
		return ""
	}
}

// ---- JSON serialization (JSON.stringify(v, null, 1)) ----

func serialize(b *strings.Builder, v interface{}, depth int) {
	switch val := v.(type) {
	case *omap:
		if len(val.keys) == 0 {
			b.WriteString("{}")
			return
		}
		b.WriteString("{\n")
		inner := depth + 1
		pad := strings.Repeat(" ", inner)
		for i, k := range val.keys {
			if i > 0 {
				b.WriteString(",\n")
			}
			b.WriteString(pad)
			writeJSONString(b, k)
			b.WriteString(": ")
			serialize(b, val.vals[i], inner)
		}
		b.WriteByte('\n')
		b.WriteString(strings.Repeat(" ", depth))
		b.WriteByte('}')
	case []interface{}:
		if len(val) == 0 {
			b.WriteString("[]")
			return
		}
		b.WriteString("[\n")
		inner := depth + 1
		pad := strings.Repeat(" ", inner)
		for i, item := range val {
			if i > 0 {
				b.WriteString(",\n")
			}
			b.WriteString(pad)
			serialize(b, item, inner)
		}
		b.WriteByte('\n')
		b.WriteString(strings.Repeat(" ", depth))
		b.WriteByte(']')
	case string:
		writeJSONString(b, val)
	case bool:
		if val {
			b.WriteString("true")
		} else {
			b.WriteString("false")
		}
	case float64:
		b.WriteString(numfmt.Format(val))
	case nil:
		b.WriteString("null")
	default:
		b.WriteString("null")
	}
}

const hexDigits = "0123456789abcdef"

func writeJSONString(b *strings.Builder, s string) {
	b.WriteByte('"')
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch c {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\n':
			b.WriteString(`\n`)
		case '\r':
			b.WriteString(`\r`)
		case '\t':
			b.WriteString(`\t`)
		case '\b':
			b.WriteString(`\b`)
		case '\f':
			b.WriteString(`\f`)
		default:
			if c < 0x20 {
				b.WriteString(`\u00`)
				b.WriteByte(hexDigits[c>>4])
				b.WriteByte(hexDigits[c&0xf])
			} else {
				b.WriteByte(c)
			}
		}
	}
	b.WriteByte('"')
}

File: python/src/gridmd/dump.py
"""Canonical model dump — the cross-language conformance contract.

Byte-identical JSON to the reference ``dumpModel`` (``js/src/dump.ts``):
``JSON.stringify(value, null, 1)`` — 1-space indent, fixed key order (insertion
order, never alphabetized except the contract-sorted arrays), shortest
round-trip numbers, and a trailing newline.
"""

from __future__ import annotations

import math
from typing import Any

from .model import CellContent, Sheet, WorkbookModel
from .numfmt import format_number
from .refs import num_to_col
from .scalar import Scalar


def _scalar_dump(s: Scalar | None) -> Any:
    if s is None:
        return None
    if s.kind == "number":
        return {"t": "n", "v": s.value}
    if s.kind == "boolean":
        return {"t": "b", "v": s.value}
    if s.kind == "error":
        return {"t": "e", "v": s.value}
    if s.kind in ("date", "time"):
        return {"t": "d", "v": s.value}
    return {"t": "s", "v": "" if s.value is None else str(s.value)}


def _content_dump(ct: CellContent) -> Any:
    if ct.rich:
        text = "".join("" if r.get("text") is None else str(r.get("text")) for r in ct.rich)
        return {"t": "rich", "v": text}
    if ct.formula is not None:
        return {"t": "f", "f": ct.formula, "cached": _scalar_dump(ct.cached), "array": ct.array_ref}
    return _scalar_dump(ct.scalar)


def _js_string(v: Any) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return format_number(v)
    return str(v)


def _dump_names(fm: dict) -> list:
    raw = fm.get("names") or []
    entries = []
    for item in raw:
        if not isinstance(item, dict):
            continue
        name = item.get("name")
        name = name if isinstance(name, str) else ""
        entry = {
            "name": name,
            "ref": item.get("ref") if item.get("ref") is not None else None,
            "formula": item.get("formula") if item.get("formula") is not None else None,
            "value": _js_string(item["value"]) if "value" in item else None,
        }
        entries.append((name, entry))
    entries.sort(key=lambda e: e[0])
    return [e[1] for e in entries]


def _dump_cells(s: Sheet) -> dict:
    content_cells = [c for c in s.cells.values() if c.content is not None]
    content_cells.sort(key=lambda c: (c.row, c.col))
    out: dict = {}
    for c in content_cells:
        out[f"{num_to_col(c.col)}{c.row}"] = _content_dump(c.content)
    return out


def _dump_merges(s: Sheet) -> list:
    refs = [f"{num_to_col(m.c1)}{m.r1}:{num_to_col(m.c2)}{m.r2}" for m in s.merges]
    refs.sort()
    return refs


def _dump_tables(s: Sheet) -> list:
    tables = sorted(s.tables, key=lambda t: t.name)
    out = []
    for t in tables:
        out.append(
            {
                "name": t.name,
                "anchor": f"{num_to_col(t.anchor.col)}{t.anchor.row}",
                "columns": list(t.columns),
                "bodyRows": t.body_rows,
                "hasTotals": bool(t.total),
            }
        )
    return out


def _cf_count(s: Sheet) -> int:
    return sum(len(rules) if isinstance(rules, list) else 0 for rules in s.cf)


def _dump_sheet(s: Sheet) -> dict:
    meta = s.meta
    hidden_raw = meta.get("hidden")
    hidden: Any = True if hidden_raw is True else "very" if hidden_raw == "very" else False
    freeze = meta.get("freeze")
    freeze = freeze if isinstance(freeze, str) else None
    protect = meta.get("protect")
    protected = bool(protect.get("enabled")) if isinstance(protect, dict) else False
    return {
        "name": s.name,
        "kind": s.kind,
        "hidden": hidden,
        "freeze": freeze,
        "protected": protected,
        "cells": _dump_cells(s),
        "merges": _dump_merges(s),
        "tables": _dump_tables(s),
        "counts": {
            "cf": _cf_count(s),
            "validations": len(s.validations),
            "notes": len(s.notes),
            "threads": len(s.threads),
            "scenarios": len(s.scenarios),
            "sparklines": len(s.sparklines),
            "charts": len(s.charts),
            "pivots": len(s.pivots),
            "slicers": len(s.slicers),
            "images": len(s.images),
            "shapes": len(s.shapes),
            "hyperlinks": len(s.hyperlinks),
        },
    }


def dump_model(model: WorkbookModel) -> str:
    fm = model.fm
    date_system = 1904 if fm.get("date-system") == 1904 else 1900
    out = {
        "gridmd": fm.get("gridmd") if _is_primitive(fm.get("gridmd")) else None,
        "title": fm.get("title") if _is_primitive(fm.get("title")) else None,
        "dateSystem": date_system,
        "names": _dump_names(fm),
        "sheets": [_dump_sheet(s) for s in model.sheets],
    }
    buf: list[str] = []
    _serialize(buf, out, 0)
    buf.append("\n")
    return "".join(buf)


def _is_primitive(v: Any) -> bool:
    return v is None or isinstance(v, (str, bool, int, float))


# ---- JSON serialization (JSON.stringify(v, null, 1)) ----
_ESCAPES = {'"': '\\"', "\\": "\\\\", "\n": "\\n", "\r": "\\r", "\t": "\\t", "\b": "\\b", "\f": "\\f"}


def _write_json_string(buf: list[str], s: str) -> None:
    buf.append('"')
    for ch in s:
        esc = _ESCAPES.get(ch)
        if esc is not None:
            buf.append(esc)
        elif ord(ch) < 0x20:
            buf.append(f"\\u{ord(ch):04x}")
        else:
            buf.append(ch)
    buf.append('"')


def _serialize(buf: list[str], v: Any, depth: int) -> None:
    if v is None:
        buf.append("null")
    elif isinstance(v, bool):
        buf.append("true" if v else "false")
    elif isinstance(v, str):
        _write_json_string(buf, v)
    elif isinstance(v, (int, float)):
        buf.append(format_number(v) if math.isfinite(float(v)) else "null")
    elif isinstance(v, dict):
        if not v:
            buf.append("{}")
            return
        buf.append("{\n")
        pad = " " * (depth + 1)
        for i, (k, val) in enumerate(v.items()):
            if i > 0:
                buf.append(",\n")
            buf.append(pad)
            _write_json_string(buf, k)
            buf.append(": ")
            _serialize(buf, val, depth + 1)
        buf.append("\n" + " " * depth + "}")
    elif isinstance(v, list):
        if not v:
            buf.append("[]")
            return
        buf.append("[\n")
        pad = " " * (depth + 1)
        for i, item in enumerate(v):
            if i > 0:
                buf.append(",\n")
            buf.append(pad)
            _serialize(buf, item, depth + 1)
        buf.append("\n" + " " * depth + "]")
    else:
        raise TypeError(f"cannot serialize {type(v).__name__} to canonical JSON")
